Support mesos fetcher

wendigo opened the pull request:
Few notes:

- there is new "fetch" field introduced in V2 API
- fetch field is serialized into CommandInfo.Uri (shared with "uris" field)
- when deserializing CommandInfo.Uri into AppDefinition both "uris" and "fetch" fields are populated
- when using V2AppDefinition, V2AppUpdate validation occurs that checks if only one of [fetch, uris]s fields are populated with data
- this change is backward-compatible, new field allows to pass known fetcher flags at the moment (executable, cached, extract)
- fetch uri is checked against known and supported protocols by the mesos fetcher (http(s), ftp(s), hdfs)

wendigo commented:
Waiting for comments :)

aquamatthias commented:
@wendigo: thanks for this PR! Will look into this!

wendigo commented:
@aquamatthias: thx!

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/Validation.scala:
```diff
@@ -91,6 +93,25 @@ object Validation {
+    new Validator[FetchUri] {
+      def apply(uri: FetchUri) = {
+        Try {
+          val value = new URL(uri.uri)
```
Simple file uris like `/etc/foo` will fail here.

wendigo replied:
Currently mesos supports three kinds of fetcher: curl, docker and hadoop (https://github.com/apache/mesos/blob/cb1d7be772d2b4864cde3620c85e67bc4330f638/src/uri/fetcher.hpp)

I don't fully understand why simple file uris should be supported if they are not supported by mesos

aquamatthias replied:
Current implementation allows simple  files, ftp, s3 etc.
There is also a plugin mechanism to extend that functionality.
So: I don't think we can validate here.

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/Validation.scala:
```diff
@@ -91,6 +93,25 @@ object Validation {
   }
 
+  def fetchUriHasSupportedProtocol: Validator[FetchUri] = {
+    lazy val supportedProtocols = Set("http", "https", "ftp", "ftps", "hdfs")
```
We tightly couple the capabilities of mesos to this validation here. 
As much as I like to fail fast, this could also be a limitation.

wendigo replied:
You are right - this will require to keep mesos/marathon capabilities in sync. I can remove this validation but failure will occur much deeper - on the mesos level

aquamatthias replied:
Please remove - see comment above.

aquamatthias commented on docs/docs/rest-api/public/api/v2/schema/AppDefinition.json:
```diff
@@ -202,6 +202,39 @@
+      "type": "array",
+      "description": "Provided URIs are passed to Mesos fetcher module and resolved in runtime.",
+      "items": {
+        "type": "object",
```
add: "additionalProperties": false to this object.

wendigo replied:
done

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala:
```diff
@@ -498,21 +512,20 @@ trait V2Formats {
       .withDefault(AppDefinition.DefaultMaxLaunchDelay.toSeconds).asSeconds ~
       (__ \ "container").readNullable[Container] ~
-      (__ \ "healthChecks").readNullable[Set[HealthCheck]].withDefault(AppDefinition.DefaultHealthChecks) ~
-      (__ \ "dependencies").readNullable[Set[PathId]].withDefault(AppDefinition.DefaultDependencies)
```
You removed dependencies?

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala:
```diff
@@ -166,6 +169,23 @@ object V2AppDefinition {
     }
   }
 
+  private def definitionContainsEitherUrisOrFetch: Validator[V2AppDefinition] = {
```
Minor: just `containsEitherUrisOrFetch`?

wendigo replied:
done

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/V2AppUpdate.scala:
```diff
@@ -119,11 +124,31 @@ case class V2AppUpdate(
   import com.wix.accord.dsl._
   import mesosphere.marathon.api.v2.Validation._
+
+  private def updateContainsEitherUrisOrFetch: Validator[V2AppUpdate] = {
```
Minor: just `containsEitherUrisOrFetch`?

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/V2AppUpdate.scala:
```diff
@@ -119,11 +124,31 @@ case class V2AppUpdate(
+        val fetch = app.fetch.getOrElse(AppDefinition.DefaultFetch)
+        val uris = app.uris.getOrElse(AppDefinition.DefaultUris)
+
+        if (fetch.nonEmpty && uris.nonEmpty) {
```
app.fetch and app.uris is an Option, so you can do:
```scala
if (app.fetch.nonEmpty && app.uris.nonEmpty) 
```
without mapping into the option.

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala:
```diff
@@ -42,6 +41,8 @@ case class V2AppDefinition(
 
     uris: Seq[String] = AppDefinition.DefaultUris,
```
Not 100% sure - but could we replace `uris` with a `def` and compute it from `fetch` (or even remove uris)?
The Format[V2AppDefinition] has to read/render uris and fetch.
Going this route, validation has to be done on the Json.reads function (see ipAddress as example).
With that, we basically replace uris as array of String with fetch as array of FetchUri's, which would make the code more clear and easy to understand.
What do you think?

aquamatthias commented:
@wendigo Thank you very much for this PR. It really looks very good!
I left a couple of comments. (biggest concern: can we get rid of uris as field in AppDefinition and still maintain the json format - see comment).
And also: were you able to test the functionality?

wendigo commented:
Hi @aquamatthias: I've refactored this change and removed uris both from V2AppDefinition and V2AppUpdate.

Field is still exposed in V2 API in "uris" field. Opposed to previous change I'm now combining uris and fetch fields together as distinct sequence of "FetchUri" objects so both fields can be used together. When using uris field "extract" param is derived from uri extension.

I will do some manual testing and will try to create more tests to cover some missing cases.

aquamatthias commented:
@wendigo thanks! and looks good. I am interested in your test results.

aquamatthias commented:
Hey @wendigo we removed V2AppDefinition in a recent commit, so we only have one AppDefinition/AppUpdate.
Can you please rebase on that?

wendigo commented:
Hey @aquamatthias - I've rebased my change, run all tests and did some manual testing against master with mesos 0.26.0 and everything works like a charm :) No regressions spotted and I can confirm that mesos fetcher is indeed receiving all the flags and caching works.

```
I0118 20:54:57.715751  3339 fetcher.cpp:422] Fetcher Info: {"cache_directory":"\/tmp\/mesos\/fetch\/slaves\/220ff644-80b7-4217-9aae-93ff230f2232-S0\/serafin","items":[{"action":"BYPASS_CACHE","uri":{"cache":true,"executable":false,"extract":true,"value":"http:\/\/allegro.pl\/pagesService\/static\/frontend-bootstrap\/styles\/frontend-bootstrap.min-30c27d54bf.css"}},{"action":"BYPASS_CACHE","uri":{"cache":false,"executable":true,"extract":true,"value":"http:\/\/ucdn.allegrostatic.pl\/headerService\/static\/images\/allegro-pl-f4a42ae095.png"}},{"action":"RETRIEVE_FROM_CACHE","cache_filename":"c2-require-44_0fb09da.js","uri":{"cache":true,"executable":false,"extract":true,"value":"http:\/\/ucdn.allegrostatic.pl\/headerService\/static\/components\/require-4400fb09da.js"}}],"sandbox_directory":"\/tmp\/mesos\/0\/slaves\/220ff644-80b7-4217-9aae-93ff230f2232-S0\/frameworks\/1a6768b9-6d23-4f49-a20b-40768260c870-0000\/executors\/pl.allegro.test_serafin.59ca7f2c-be1d-11e5-88d2-080027368330\/runs\/03e31717-10e5-4391-a8f5-72f4959278d7","user":"serafin"}
I0118 20:54:57.718894  3339 fetcher.cpp:377] Fetching URI 'http://allegro.pl/pagesService/static/frontend-bootstrap/styles/frontend-bootstrap.min-30c27d54bf.css'
I0118 20:54:57.718919  3339 fetcher.cpp:248] Fetching directly into the sandbox directory
I0118 20:54:57.718938  3339 fetcher.cpp:185] Fetching URI 'http://allegro.pl/pagesService/static/frontend-bootstrap/styles/frontend-bootstrap.min-30c27d54bf.css'
I0118 20:54:57.718956  3339 fetcher.cpp:132] Downloading resource from 'http://allegro.pl/pagesService/static/frontend-bootstrap/styles/frontend-bootstrap.min-30c27d54bf.css' to '/tmp/mesos/0/slaves/220ff644-80b7-4217-9aae-93ff230f2232-S0/frameworks/1a6768b9-6d23-4f49-a20b-40768260c870-0000/executors/pl.allegro.test_serafin.59ca7f2c-be1d-11e5-88d2-080027368330/runs/03e31717-10e5-4391-a8f5-72f4959278d7/frontend-bootstrap.min-30c27d54bf.css'
W0118 20:54:58.020812  3339 fetcher.cpp:270] Copying instead of extracting resource from URI with 'extract' flag, because it does not seem to be an archive: http://allegro.pl/pagesService/static/frontend-bootstrap/styles/frontend-bootstrap.min-30c27d54bf.css
I0118 20:54:58.020870  3339 fetcher.cpp:454] Fetched 'http://allegro.pl/pagesService/static/frontend-bootstrap/styles/frontend-bootstrap.min-30c27d54bf.css' to '/tmp/mesos/0/slaves/220ff644-80b7-4217-9aae-93ff230f2232-S0/frameworks/1a6768b9-6d23-4f49-a20b-40768260c870-0000/executors/pl.allegro.test_serafin.59ca7f2c-be1d-11e5-88d2-080027368330/runs/03e31717-10e5-4391-a8f5-72f4959278d7/frontend-bootstrap.min-30c27d54bf.css'
I0118 20:54:58.020889  3339 fetcher.cpp:377] Fetching URI 'http://ucdn.allegrostatic.pl/headerService/static/images/allegro-pl-f4a42ae095.png'
I0118 20:54:58.020895  3339 fetcher.cpp:248] Fetching directly into the sandbox directory
I0118 20:54:58.020912  3339 fetcher.cpp:185] Fetching URI 'http://ucdn.allegrostatic.pl/headerService/static/images/allegro-pl-f4a42ae095.png'
I0118 20:54:58.020927  3339 fetcher.cpp:132] Downloading resource from 'http://ucdn.allegrostatic.pl/headerService/static/images/allegro-pl-f4a42ae095.png' to '/tmp/mesos/0/slaves/220ff644-80b7-4217-9aae-93ff230f2232-S0/frameworks/1a6768b9-6d23-4f49-a20b-40768260c870-0000/executors/pl.allegro.test_serafin.59ca7f2c-be1d-11e5-88d2-080027368330/runs/03e31717-10e5-4391-a8f5-72f4959278d7/allegro-pl-f4a42ae095.png'
I0118 20:54:58.128809  3339 fetcher.cpp:454] Fetched 'http://ucdn.allegrostatic.pl/headerService/static/images/allegro-pl-f4a42ae095.png' to '/tmp/mesos/0/slaves/220ff644-80b7-4217-9aae-93ff230f2232-S0/frameworks/1a6768b9-6d23-4f49-a20b-40768260c870-0000/executors/pl.allegro.test_serafin.59ca7f2c-be1d-11e5-88d2-080027368330/runs/03e31717-10e5-4391-a8f5-72f4959278d7/allegro-pl-f4a42ae095.png'
I0118 20:54:58.128849  3339 fetcher.cpp:377] Fetching URI 'http://ucdn.allegrostatic.pl/headerService/static/components/require-4400fb09da.js'
I0118 20:54:58.128856  3339 fetcher.cpp:287] Fetching from cache
W0118 20:54:58.128883  3339 fetcher.cpp:315] Copying instead of extracting resource from URI with 'extract' flag, because it does not seem to be an archive: http://ucdn.allegrostatic.pl/headerService/static/components/require-4400fb09da.js
I0118 20:54:58.128895  3339 fetcher.cpp:165] Copying resource with command:cp '/tmp/mesos/fetch/slaves/220ff644-80b7-4217-9aae-93ff230f2232-S0/serafin/c2-require-44_0fb09da.js' '/tmp/mesos/0/slaves/220ff644-80b7-4217-9aae-93ff230f2232-S0/frameworks/1a6768b9-6d23-4f49-a20b-40768260c870-0000/executors/pl.allegro.test_serafin.59ca7f2c-be1d-11e5-88d2-080027368330/runs/03e31717-10e5-4391-a8f5-72f4959278d7/require-4400fb09da.js'
I0118 20:54:58.132892  3339 fetcher.cpp:454] Fetched 'http://ucdn.allegrostatic.pl/headerService/static/components/require-4400fb09da.js' to '/tmp/mesos/0/slaves/220ff644-80b7-4217-9aae-93ff230f2232-S0/frameworks/1a6768b9-6d23-4f49-a20b-40768260c870-0000/executors/pl.allegro.test_serafin.59ca7f2c-be1d-11e5-88d2-080027368330/runs/03e31717-10e5-4391-a8f5-72f4959278d7/require-4400fb09da.js'
```

aquamatthias commented:
Great and Thanks! Can you remove the protocol validation?

wendigo commented:
Done :)

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/Validation.scala:
```diff
@@ -92,6 +93,17 @@ object Validation {
+    new Validator[FetchUri] {
+      def apply(uri: FetchUri) = {
+        Try {
+          new URL(uri.uri)
```
You validate a URI by parsing a URL (and to make it more complicated: a java  based URL, which only knows about specific protocols and schemes).
A valid URI is: `/etc/docker` or `hdfs://namenode-host-name:web-interface-port/data/file-path`
which both fail on my machine.
You could use to parse a URI (http://www.ietf.org/rfc/rfc2396.txt) instead.

aquamatthias replied:
See fetcher cache documentation here: http://mesos.apache.org/documentation/latest/fetcher/
Excerpt:

> The Mesos fetcher can copy files from a local filesytem and it also natively supports the HTTP, HTTPS, FTP and FTPS protocols. If the requested URI is based on some other protocol, then the fetcher tries to utilise a local Hadoop client and hence supports any protocol supported by the Hadoop client, e.g., HDFS, S3. See the slave configuration documentation for how to configure the slave with a path to the Hadoop client.

wendigo replied:
Done.

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala:
```diff
@@ -505,21 +524,21 @@ trait V2Formats {
-        backoffFactor = backoffFactor, maxLaunchDelay = maxLaunchDelay, container = container, healthChecks = checks,
-        dependencies = dependencies
+        mem = mem, disk = disk, executor = executor, constraints = constraints,
+        fetch = mergeFetchWithUris(fetch, uris), storeUrls = storeUrls, ports = Nil, requirePorts = requirePorts,
```
Does this allow specifying `uris` and `fetch`. We should enforce the usage of one or the other but not to mix (for reading). We have to fill both values for backward compatibility (writing).

wendigo replied:
Yes it does but I don't think it's really a problem. Mesos handles duplicated fetch uris without a problem

aquamatthias replied:
I am not so much concerned about mesos, but more about the user. 
What is the result of:
```json
{
  "uris": ["/my/file.sh.tar"],
  "fetch": { "uri": "/my/file.sh.tar", "extract": false }
}
```
It is just hard to guess. Can we allow either one or the other? See the filter validation in line 559.

wendigo replied:
Would you like to filter conflicting fetch uris or disallow merging uris with fetch?

aquamatthias replied:
I would disallow defining `uris` and `fetch` at the same time.
Idea: deprecate `uris`

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/Validation.scala:
```diff
@@ -92,6 +92,17 @@ object Validation {
+  def fetchUriIsValid: Validator[FetchUri] = {
+    new Validator[FetchUri] {
+      def apply(uri: FetchUri) = {
+        Try {
```
replace with an ordinary try/catch - since you don't use the Try Monad at all.

wendigo replied:
Done

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala:
```diff
@@ -697,60 +719,75 @@ trait V2Formats {
-        }
+  private[this] def addUrisToFetch(fetch: Option[Seq[FetchUri]], uris: Option[Seq[String]]): Option[Seq[FetchUri]] = {
+    if (uris.nonEmpty) {
+      val converted = uris.getOrElse(Seq.empty).map(uri => FetchUri(uri = uri, extract = FetchUri.isExtract(uri)))
```
Is that still needed?
Should be `eitherUriOrFetch: Option[Seq[FetchUri]]` ?

wendigo replied:
I've changed merge strategy in Formats to either allow fetch or uris field and lot's of unrelated tests failed :( I think it's due to json being read/write in more complex roundtrips

aquamatthias replied:
probably misleading comment - sorry.
I don't think you have to merge here - correct?
Either map defined `uris` or use `fetch` as is.

wendigo replied:
Yes, but on read path both fields are populated

aquamatthias replied:
Read: either `uris` or `fetch` --> will land in `fetch`
Write: `fetch` will populate `uris` and `fetch`
correct?

wendigo replied:
Ok, so it works as you've said - either fetch or uris field can be populated on read - both are populated on read

aquamatthias commented on src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala:
```diff
@@ -697,60 +719,75 @@ trait V2Formats {
+    (__ \ "disk").readNullable[JDouble] ~
+    (__ \ "executor").readNullable[String](Reads.pattern("^(//cmd)|(/?[^/]+(/[^/]+)*)|$".r)) ~
+    (__ \ "constraints").readNullable[Set[Constraint]] ~
+    (__ \ "uris").readNullable[Seq[String]] ~
```
Should also make sure, either uris or fetch is defined.

wendigo replied:
And what about this kind of situation:

```
{
  "fetch": [{ "uri": "/my/file.sh.tar", "extract": false },{ "uri": "/my/file.sh.tar", "extract": true }]
}
```

wendigo commented:
Ok so the build failure is unrelated to my change. Can you rerun it?

wendigo commented:
Now I cannot build it locally: 
```
[error] Server access Error: Host is down url=http://downloads.mesosphere.io/maven/mesosphere/marathon/ui/0.15.0-SNAPSHOT/maven-metadata.xml
[error] Server access Error: Host is down url=http://downloads.mesosphere.io/maven/mesosphere/marathon/ui/0.15.0-SNAPSHOT/ui-0.15.0-SNAPSHOT.pom
[error] Server access Error: Host is down url=http://repo.typesafe.com/typesafe/releases/mesosphere/marathon/ui/0.15.0-SNAPSHOT/maven-metadata.xml
[error] Server access Error: Host is down url=http://repo.typesafe.com/typesafe/releases/mesosphere/marathon/ui/0.15.0-SNAPSHOT/ui-0.15.0-SNAPSHOT.pom
``` 
during ```sbt assembly```

aquamatthias commented:
Hey @wendigo - thanks for incorporating the comments. 
The typesafe repository is down - that is the reason for the sbt error.

wendigo commented:
Not a problem :) If you have any other thoughts - let me know.

aquamatthias commented:
:100:  Thank you very much!

wendigo commented:
Uff :) Thanks for valuable insight and patience!
